src/qe/mbp/mbp_arrays_tg.cpp

```diff
@@ -0,0 +1,367 @@
+                if (m_mdl.are_equal(j, i)) {
+                    in = true;
+                    //save for later
+                    eq_index = i;
+                    break;
+                }
+                else deq.push_back(i);
+            }
+        }
+        if (in) {
+            SASSERT(m_mdl.are_equal(j, eq_index));
+            peq p_new = mk_peq(to_app(p.lhs())->get_arg(0), p.rhs(), indices);
+            m_tg.add_eq(j, eq_index);
+            expr_ref p_new_expr(m);
+            p_new_expr = is_neg ? m.mk_not(p_new.mk_peq()) : p_new.mk_peq();
+            m_tg.add_lit(p_new_expr);
+            m_tg.add_eq(p_new_expr, p.mk_peq());
+            return;
+        }
+        for(expr* d : deq) {
+            m_tg.add_deq(j, d);
+        }
+        expr_ref_vector setOne(m);
+        setOne.push_back(j);
+        indices.push_back(setOne);
+        peq p_new = mk_peq(to_app(p.lhs())->get_arg(0), p.rhs(), indices);
+        expr* args[2] = {p.rhs(), j};
+        expr_ref rd(m_array_util.mk_select(2, args), m);
+        if (!is_neg) {
+            m_tg.add_lit(p_new.mk_peq());
+            m_tg.add_eq(rd, elem);
+            m_tg.add_eq(p.mk_peq(), p_new.mk_peq());
+        }
+        else {
+            SASSERT(m_mdl.is_false(p_new.mk_peq()) || !m_mdl.are_equal(rd, elem));
+            if (m_mdl.is_false(p_new.mk_peq())) {
+                expr_ref npeq(mk_not(p_new.mk_peq()), m);
+                m_tg.add_lit(npeq);
+                m_tg.add_eq(p.mk_peq(), p_new.mk_peq());
+            }
+            if (!m_mdl.are_equal(rd, elem)) {
+                m_tg.add_deq(rd, elem);
+            }
+        }
+    }
+
+    void add_rdVar(expr* rd) {
+        //do not assign new variable if rd is already equal to a value
+        if (m_tg.has_val_in_class(rd)) return;
+        TRACE("mbp_tg", tout << "applying add_rdVar on " << expr_ref(rd, m););
+        app_ref u = new_var(to_app(rd)->get_sort(), m);
+        m_new_vars.push_back(u);
+        m_tg.add_var(u);
+        m_tg.add_eq(u, rd);
+        m_mdl.register_decl(u->get_decl(), m_mdl(rd));
+    }
+
+    void elimeq(peq p) {
+        TRACE("mbp_tg", tout << "applying elimeq on " << expr_ref(p.mk_peq(), m););
+        app_ref_vector aux_consts(m);
+        expr_ref eq(m);
+        expr_ref sel(m);
+        eq = p.mk_eq(aux_consts, true);
+        vector<expr_ref_vector> indices;
+        p.get_diff_indices(indices);
+        vector<expr_ref_vector>::iterator itr = indices.begin();
+        unsigned i = 0;
+        for(app* a : aux_consts) {
+            m_new_vars.push_back(a);
+            m_tg.add_var(a);
+            auto const& indx =  std::next(itr, i);
+            SASSERT(indx->size() == 1);
+            expr *args[2] = {to_app(p.lhs()), to_app(indx->get(0))};
+            sel = m_array_util.mk_select(2, args);
+            m_mdl.register_decl(a->get_decl(), m_mdl(sel));
+            i++;
+        }
+        m_tg.add_lit(eq);
+        m_tg.add_lit(m.mk_true());
```

Comment:
this will try to internalize true many times, can it be optimized? In the net call using `add_eq` a term for true will also be created